src/auth.py

```diff
@@ -2,35 +2,143 @@
 import os
 import uuid
 
-import src.constants.api_keys as api_key
-from src.config import UNAUTHORIZED_REQUEST_CODE, BAD_REQUEST_CODE, TOKEN_ENV_VAR
+import jsonschema
+
+import src.constants.keys as key
+import src.constants.constants as consts
+from src.config import UNAUTHORIZED_REQUEST_CODE, BAD_REQUEST_CODE, CLUSTER_TOKEN_ENV_VAR
+from src.database import DatabaseResolver
 
 logger = logging.getLogger()
 
 
-def initialize_token():
-    if os.environ.get(TOKEN_ENV_VAR):
-        logger.info("using authentication token from environment")
-        token = os.environ.get(TOKEN_ENV_VAR)
+def initialize_cluster_token():
+    if os.environ.get(CLUSTER_TOKEN_ENV_VAR):
+        logger.info("using cluster token from environment")
+        token = os.environ.get(CLUSTER_TOKEN_ENV_VAR)
     else:
-        logger.info("generating authentication token")
+        logger.info("generating cluster token")
         token = str(uuid.uuid4())
-        logger.info("authentication token is {}".format(token))
+        logger.info("cluster token is {}".format(token))
     return token
 
 
-def authenticate(func):
+def configure_course_tokens(db_resolver, course_tokens):
```

Comment:
The courses and their tokens have a many-to-many relationship. A course can own multiple tokens (ex a course might have different systems pinging the API and might want to have independent tokens for them) and a token can be owned by multiple courses, for instance, two courses "cs241" and "cs241-honors" should be able to share the same token if they want.

This function aims to understand that relationship and feed it into the DB which is why I think it might be confusing. An example config can look like:
```
{  
   "tokens":{  
      "token1":"abcd",
      "token2":"efgh"
   },
   "courses":{  
      "cs241-honors":[  
         "token1"
      ],
      "cs241":[  
         "token1",
         "token2"
      ]
   }
}
```

I added this to provide maximal flexibility to courses to chose their tokens the way they want. Let me know if this is a fair design or not. We could always reduce to a one-to-many model where each course has a list of tokens which would simplify things.

As for resolving this, I could add more docs.

Reply:
This seems a bit over-engineered to me, and that's probably why it's confusing. At some point offering too much flexibility just makes things hard to maintain, and I think this might be one of those cases.

Why not just have courses, and each course has a list of tokens? I can't really imagine a realistic case where someone would say they wish they had a many-to-many relationship.